### ISSUE TITLE
 Cannot read property 'externalModuleIndicator' of undefined error occurred when trying to compile 

### ISSUE THREAD

**Issue by aiiak** [Microsoft/TypeScript#15725]
**TypeScript Version:**  2.3.2
Getting following error when trying to compile my solution  with  typescript 2.3.2
```
TypeError: Cannot read property 'externalModuleIndicator' of undefined
    at isBlockScopedNameDeclaredBeforeUse (C:\Users\Alexander.Yakovlev\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:21253:87)
    at checkResolvedBlockScopedVariable (C:\Users\Alexander.Yakovlev\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:21613:57)
    at resolveName (C:\Users\Alexander.Yakovlev\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:21502:25)
    at getGlobalSymbol (C:\Users\Alexander.Yakovlev\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:26100:20)
    at getGlobalValueSymbol (C:\Users\Alexander.Yakovlev\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:26094:20)
    at getGlobalPromiseConstructorSymbol (C:\Users\Alexander.Yakovlev\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:26122:104)
    at createPromiseReturnType (C:\Users\Alexander.Yakovlev\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:32996:23)
    at getReturnTypeFromBody (C:\Users\Alexander.Yakovlev\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:33037:31)
    at getReturnTypeOfSignature (C:\Users\Alexander.Yakovlev\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:25737:28)
    at checkCallExpression (C:\Users\Alexander.Yakovlev\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:32837:20)
```
Doesn't have any problems with ts@2.2.1 

It looks like  somehow caused by  Promise< T > implementation along with   compilerOptions.target = "es5".  If we remove  Promise<T> from solution, error, that  i provided above  didn't occurred, and we got  
```
error TS2705: An async function or method in ES5/ES3 requires the 'Promise' constructor.
```

**Comment by RyanCavanaugh** [comment]
We need a way to reproduce this error

**Comment by aiiak** [comment]
Hi, Ryan. 
I cannot provide you all my solution and cannot reproduce  problem in a smaller one , but can you please look at call stack, cause it looks something wrong here. 

In tsc.js:26100 there is getGlobalSymbol call with undefined as first param
```typescript
function getGlobalSymbol(name, meaning, diagnostic) {
    return resolveName(undefined, name, meaning, diagnostic, name);
}
```
In resolveName (tsc.js:21502)  it assigned to errorLocation and later pass to checkResolvedBlockScopedVariable
```typescript 
function resolveName(location, name, meaning, nameNotFoundMessage, nameArg) {
    ...
    var errorLocation = location; 
    ...
    checkResolvedBlockScopedVariable(exportOrLocalSymbol, errorLocation);
    ...
}
```
Then,  in checkResolvedBlockScopedVariable, there is isBlockScopedNameDeclaredBeforeUse(declaration, errorLocation) call
```typescript
function checkResolvedBlockScopedVariable(result, errorLocation) {  
    ...
    if (!ts.isInAmbientContext(declaration) && !isBlockScopedNameDeclaredBeforeUse(declaration, errorLocation)) 
    ...
}
```
i guess there is a problem, cause isBlockScopedNameDeclaredBeforeUse is't looks like a function that is expected second parameter could be undefined
```typescript
function isBlockScopedNameDeclaredBeforeUse(declaration, usage) {
    var declarationFile = ts.getSourceFileOfNode(declaration);
    var useFile = ts.getSourceFileOfNode(usage);
    if (declarationFile !== useFile) {
        if ((modulekind && (declarationFile.externalModuleIndicator || useFile.externalModuleIndicator))
...
```
if usage = undefined, then useFile will be  undefined also and useFile.externalModuleIndicator will rise "Cannot read property 'externalModuleIndicator' of undefined" exceptrion,   as happens in my case. 
Thanks.

**Comment by sandersn** [comment]
@aiiak are you building with any tools like webpack or gulp that might alter typescript's global environment? Webpack messes up the global environment when typechecking is turned off, for example.

**Comment by rbuckton** [comment]
Here's a simple repro:
```ts
// @target: es5
class Promise<T> {}
async function foo() {}
const x = foo();
```

**Comment by sandersn** [comment]
Fix is up at #16375

**Comment by pocesar** [comment]
happening again on `typescript@2.6.0-dev.20170910`
